Issue3618 air filter

JayHuLBL opened the pull request:
This closes #3618.

JayHuLBL commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,315 @@
+<p>
+The input boolean flag, <code>uRep</code>, triggers the filter replacement.
+When <code>uRep</code> changes from <code>false</code> to <code>true</code>, the
+mass of the captured contaminant becomes zero.
```
Does it become zero, or `mCon_reset`?

JayHuLBL commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,315 @@
+<p>
+A warning will be triggered when the captured contaminant mass becomes greater than the
+maximum contaminant mass (<code>mCon_nominal</code>).
+In addition, the <code>extraPropertiesNames</code> has to be defined in the medium model.
```
Can we implement a warning? See if `Buildings.Fluid.Sensors.TraceSubstances` gives any hint.

JayHuLBL commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,315 @@
+    "Filter efficiency curve"
+    annotation (Dialog(group="Efficiency"));
+  parameter Real b(
+    final min = 1 + 1E-3)
```
why should we set the `min=1 + 1E-3`, rather than `min=1`? I see in `Buildings.Fluid.AirFilters.BaseClasses.FlowCoefficientCorrection` it says that the `b` has to be greater than 1. Should we have implement a warning?

JayHuLBL commented:
@SenHuang19 See inline comments.

SenHuang19 commented:
- [ ] Add a data record for the filter efficiency rather than asking users to do a polynomial fit.
- [ ] Change the <code>Buildings.Fluid.AirFilters.Empirical.Examples</code> into <code>Buildings.Fluid.AirFilters.Empirical.Validation</code>
- [ ] Add <code>Buildings.Fluid.AirFilters.Empirical.Examples</code> where a realistic example will be built based on
<code>Buildings.Examples.VAVCO2.VAVSystemCTControl</code>
- [ ] Eliminate typos and improve the model doc

mwetter commented:
This is ready for final review by Jianjun.

JayHuLBL commented on Buildings/Fluid/AirFilters/BaseClasses/FiltrationEfficiency.mo:
```diff
@@ -0,0 +1,89 @@
+    "Filtration efficiency of each contaminant"
+    annotation (Placement(transformation(extent={{100,-80},{140,-40}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput rat(
+    final unit="1",
```
In this class, the `rat` is a scale variable, while in class `Buildings.Fluid.AirFilters.BaseClasses.Characteristics.FiltrationEfficiencyParameters`, the `rat` is a vector. Are they same of different?
Through commit [e746684](https://github.com/lbl-srg/modelica-buildings/pull/3676/commits/e74668435a771556c67b852a56248f6defe49d3f), I removed `each` declaration. Is it a correct change? As in Modelica, `each` should only be used by the vector variable declaration.

SenHuang19 replied:
<code>Buildings.Fluid.AirFilters.BaseClasses.FilterationEfficiency</code> defines <code>rat</code> while in <code>Buildings.Fluid.AirFilters.BaseClasses.Characteristics.FiltrationEfficiencyParameters</code>, we declare a list of <code>rat</code> and each <code>rat</code> corresponds to a specific type of containment. Should I rename the <code>rat</code> in <code>Buildings.Fluid.AirFilters.BaseClasses.Characteristics.FiltrationEfficiencyParameters</code> to avoid confusions?

JayHuLBL commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,313 @@
+</li>
+<li>
+how the filtration efficiency changes along with the contaminant accumulation,
+specific to each type.
```
each type of what? is it each type of the contaminant? I wonder if this bullet refers to the `per.filEffPar`?

SenHuang19 replied:
each type of  containment, yes, it refers to <code>per.filEffPar</code>. I will improve the whole document section to avoid confusion.

JayHuLBL reviewed (commented):
@SenHuang19 See inline comments.

SenHuang19 commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,313 @@
+the types of contaminants can be captured by the filter;
+</li>
+<li>
+how the flow coefficient changes as contaminants build up;
```
Yes, will try to improve this.

JayHuLBL commented on Buildings/Fluid/AirFilters/Empirical.mo:
```diff
@@ -0,0 +1,304 @@
+          textColor={0,0,255},
+          textString="%name")}),
+Diagram(coordinateSystem(preserveAspectRatio=false)),
+Documentation(info="<html>
```
@SenHuang19 
Do we have any reference for the development? If yes, it's better to add it in the documentation.

JayHuLBL commented:
@SenHuang19 
In commit 6c00d23, I did the following improvements:
- [x] renamed the input `uRep` to `uRes` and changed its comment,
- [x] added the output `yRep`, a boolean output to flag if the filter should be replaced,
- [x] add the filter initial contaminant mass `mCon_reset` to the performance record.

Would you please check if the changes are good for you?

SenHuang19 commented:
> @SenHuang19 In commit [6c00d23](https://github.com/lbl-srg/modelica-buildings/commit/6c00d2389b074a5a5ff84ca77155ee442c9a0f3f), I did the following improvements:
> 
> * [x]  renamed the input `uRep` to `uRes` and changed its comment,
> * [x]  added the output `yRep`, a boolean output to flag if the filter should be replaced,
> * [x]  add the filter initial contaminant mass `mCon_reset` to the performance record.
> 
> Would you please check if the changes are good for you?

@JayHuLBL The changes look good to me— nice catch on adding <code>Buildings.Controls.OBC.CDL.Logical.Edge</code>. The only thing I wanted to ask about is the comment for <code>uRes</code>:
> `True:` the filter has been replaced and reset the accumulation

This sounds like it might be describing a state variable. Could you clarify?

JayHuLBL commented:
@SenHuang19 In your original implementation, when the captured contaminant is more than the nominal contaminant mass, it will generate a warning message. This implies that the filter would not be replaced immediately.

I thus changed the `uRep` to `uRes`, with the comment shows that the filter has been replaced and thus the accumulation calculation should be reset.

SenHuang19 commented:
> @SenHuang19 In your original implementation, when the captured contaminant is more than the nominal contaminant mass, it will generate a warning message. This implies that the filter would not be replaced immediately.
> 
> I thus changed the `uRep` to `uRes`, with the comment shows that the filter has been replaced and thus the accumulation calculation should be reset.

Thank @JayHuLBL  for the detailed explanation. What I meant to say is that `uRes` is an input, but the comment makes it sound like it's a state variable indicating whether the filter has been replaced

SenHuang19 commented on Buildings/Fluid/AirFilters/Examples/SpaceCooling.mo:
```diff
@@ -0,0 +1,332 @@
+</li>
+</ul>
+<p>
+The expected output are:
```
I'll convert them into different units and open a PR soon.

mwetter commented:
@JayHuLBL @SenHuang19 : What remains to be done to finish this model?

JayHuLBL commented:
@SenHuang19 The PR looks good to me now.

mwetter commented:
@JayHuLBL : After your review, the PR https://github.com/lbl-srg/modelica-buildings/pull/4397 was created. Does https://github.com/lbl-srg/modelica-buildings/pull/4397 still need to be merged?

JayHuLBL commented:
Close this PR and then open a new PR to see if it will reset the CI tests.
